### PR TITLE
8341784: Refactor TypeVect to use a BasicType instead of a const Type*

### PR DESCRIPTION
**Pull request by merykitty**
Hi,

This patch refactors `TypeVect` to use a `BasicType` instead of a `const Type*` as our current implementation. This conveys the element information in a clearer and safer manner.

Please take a look and leave your reviews,
Thanks a lot.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8341784](https://bugs.openjdk.org/browse/JDK-8341784): Refactor TypeVect to use a BasicType instead of a const Type* (**Enhancement** - P4)

### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)
 * [Jasmine Karthikeyan](https://openjdk.org/census#jkarthikeyan) (@jaskarth - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21414/head:pull/21414` \
`$ git checkout pull/21414`

Update a local copy of the PR: \
`$ git checkout pull/21414` \
`$ git pull https://git.openjdk.org/jdk.git pull/21414/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21414`

View PR using the GUI difftool: \
`$ git pr show -t 21414`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21414.diff">https://git.openjdk.org/jdk/pull/21414.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21414#issuecomment-2400696880)

### CODE REVIEW

**Review by vnkozlov — COMMENTED** [review]
Looks reasonable. Need to test it internally.

**Comment by merykitty** [comment]
@vnkozlov Thanks for your reviews and testings, the latest commit addresses your concern, as well as contains some minor style changes and a removal of switch case duplication.

**Review by vnkozlov — APPROVED** [review]
Good.

You need second review because change is not trivial.

File: src/hotspot/share/opto/vectorIntrinsics.cpp [hunk 1]
```diff
@@ -600,16 +599,15 @@ bool LibraryCallKit::inline_vector_shuffle_iota() {
     return false;
   }
 
-  const Type * type_bt = Type::get_const_basic_type(elem_bt);
-  const TypeVect * vt  = TypeVect::make(type_bt, num_elem);
+  const TypeVect * vt  = TypeVect::make(elem_bt, num_elem);
```
Comment (jaskarth) [inline src/hotspot/share/opto/vectorIntrinsics.cpp:602]:
```suggestion
  const TypeVect* vt  = TypeVect::make(elem_bt, num_elem);
```

File: src/hotspot/share/opto/vectorIntrinsics.cpp [hunk 2]
```diff
@@ -2202,14 +2199,13 @@ bool LibraryCallKit::inline_vector_select_from() {
   }
 
   // cast index vector from elem_bt vector to byte vector
-  const Type * byte_bt = Type::get_const_basic_type(T_BYTE);
-  const TypeVect * byte_vt  = TypeVect::make(byte_bt, num_elem);
+  const TypeVect * byte_vt  = TypeVect::make(T_BYTE, num_elem);
```
Comment (jaskarth) [inline src/hotspot/share/opto/vectorIntrinsics.cpp:2202]:
```suggestion
  const TypeVect* byte_vt  = TypeVect::make(T_BYTE, num_elem);
```

File: src/hotspot/share/opto/vectorIntrinsics.cpp [hunk 3]
```diff
@@ -618,12 +616,12 @@ bool LibraryCallKit::inline_vector_shuffle_iota() {
...
 
   if (!start_val->is_con() || start_val->get_con() != 0) {
-    Node* bcast_start    = gvn().transform(VectorNode::scalar2vector(start, num_elem, type_bt));
+    Node* bcast_start    = gvn().transform(VectorNode::scalar2vector(start, num_elem, elem_bt));
     res = gvn().transform(VectorNode::make(Op_AddVB, res, bcast_start, vt));
   }
 
   Node * mod_val = gvn().makecon(TypeInt::make(num_elem-1));
-  Node * bcast_mod  = gvn().transform(VectorNode::scalar2vector(mod_val, num_elem, type_bt));
+  Node * bcast_mod  = gvn().transform(VectorNode::scalar2vector(mod_val, num_elem, elem_bt));
```
Comment (jaskarth) [inline src/hotspot/share/opto/vectorIntrinsics.cpp:624]:
```suggestion
  Node* bcast_mod  = gvn().transform(VectorNode::scalar2vector(mod_val, num_elem, elem_bt));
```

File: src/hotspot/share/opto/type.cpp [hunk 4]
```diff
@@ -2482,58 +2482,59 @@ bool TypeAry::ary_must_be_exact() const {
...
-    return make(elem, length);
+    return make(elem_bt, length);
   }
 }
 
 //------------------------------meet-------------------------------------------
-// Compute the MEET of two types.  It returns a new Type object.
-const Type *TypeVect::xmeet( const Type *t ) const {
+// Compute the MEET of two types. Since each TypeVect is the only instance of
+// its species, meetting often returns itself
```
Comment (jaskarth) [inline src/hotspot/share/opto/type.cpp:2532]:
```suggestion
// its species, meeting often returns itself.
```

**Review by jaskarth — APPROVED** [review]
I think this is a really nice cleanup, it makes working with vectors more clear. I've added some minor stylistic fixes you could do since you're already changing these lines.

**Comment by merykitty** [comment]
@jaskarth Nice suggestions, I have reviewed the patch and done similar changes to nearby lines.

**Review by jaskarth — APPROVED** [review]
Thanks, looks good to me!

**Comment by merykitty** [comment]
@vnkozlov Could you re-review this, please, it seems required.

**Review by vnkozlov — APPROVED** [review]
Update is good.

**Comment by merykitty** [comment]
Thanks a lot for your reviews
/integrate

**Comment by jatin-bhateja** [comment]
Hi @merykitty , LGTM.

Best Regards.

**Comment by merykitty** [comment]
@jatin-bhateja Thanks a lot for your reviews.

**Comment by eme64** [comment]
@merykitty I was out on vacation and missed this patch. Can you tell me why the `long` type has the letter `J` and not `L` in the vector type?

**Comment by eme64** [comment]
Ah, I guess that is generally the letter we use when passing longs, like in method signatures. And that is because `L` is already taken for objects. Oh well, makes sense now :laughing:

**Comment by PaulSandoz** [comment]
> Ah, I guess that is generally the letter we use when passing longs, like in method signatures. And that is because `L` is already taken for objects. Oh well, makes sense now 😆

Welcome to the in-between world of Java and the JVM :-)

**Pull request closed without merging**